Add Azure Web Service docs (#299)
signalfx/signalfx-dotnet-tracing: docs/azure.md

> # Azure services instrumentation guide

Reviewer: ```suggestion
# Azure instrumentation guide
```

WDYT? to keep in sync with the header from `README.md`

> # Azure services instrumentation guide
> 
> ## App Service

Reviewer: ```suggestion
# Instrument Azure App Service
```

to have the same title as the header in `README.md`

1. Choose your App Service.
2. Navigate to `Development Tools` > `Extensions`.

> 3. Find and install `SignalFx .NET Tracing` extension.

Reviewer: ```suggestion
3. Find and install the `SignalFx .NET Tracing` extension.
```

4. Navigate to `Settings` > `Configuration`.

> 5. Add `New application setting` to configure receiver. See [advanced-config.md](advanced-config.md#exporter-settings) exporter settings section for more details.

Reviewer 2: Could you please be more specific which environmental variables are mandatory?
In previous document three of them was marked in this way `SIGNALFX_SERVICE_NAME`, `SIGNALFX_ACCESS_TOKEN` and  `SIGNALFX_ENDPOINT_URL`.

Reviewer: And probably it is worth saying that `SIGNALFX_ACCESS_TOKEN` and `SIGNALFX_ENDPOINT_URL` should be set in a way to send the telemetry directly to the Splunk Observability Cloud.

Reviewer: Also, I think a step like `Restart your App Service.` is missing at the end.

Reviewer 2: As I remember, `SIGNALFX_ACCESS_TOKEN` is mandatory. Otherwise, profiling for Azure App Service won't work.
We can change it - requierments from upstream.

Reviewer 3: @RassK Include setting `SIGNALFX_AZURE_APP_SERVICES` to true.

Author: SIGNALFX_ACCESS_TOKEN and SIGNALFX_ENDPOINT_URL are the bare minimum to make it work. 
SIGNALFX_AZURE_APP_SERVICES is setup by the extension already.

Applying settings will restart the service anyway (at least it did for me).

Reviewer 3: > SIGNALFX_AZURE_APP_SERVICES is setup by the extension already.

ops, forgot about that...

Author: > > SIGNALFX_AZURE_APP_SERVICES is setup by the extension already.
> 
> ops, forgot about that...

@pjanotti btw, just discovered that `SIGNALFX_AZURE_APP_SERVICES=1` is effectively disabling WebJobs instrumentation by mistakenly thinking it's infrastructure service.  Also non of the ENVs set by the extension are effective for WebJobs. By manually adding envs and `SIGNALFX_AZURE_APP_SERVICES=0` at least profiler attaches.

Reviewer 3: @RassK I don't understand why this happens, do you know the details?

Author: @pjanotti I think it's because Webjobs are executed in Kudu scm context (has it's own w3wp process), while actual web service generates second w3wp process. 
https://github.com/signalfx/signalfx-dotnet-tracing/blob/next-ver/tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp#L123

![image](https://user-images.githubusercontent.com/4929112/148773216-486d6f4c-49da-42c2-a385-fa2ac314cf4b.png)

Reviewer 3: @RassK I'm not that familiar with WebJobs, according to https://docs.microsoft.com/en-us/azure/app-service/webjobs-create#acceptablefiles it seems that one has to create the job as an exe to be something that we may want to instrument. It seems a very special case, at first, my instinct tells me to keep it as it is.

Author: @pjanotti yep, a console with generic host + webjobs sdk. We can leave it as is, just document that all env's must be set manually and make sure azure one is set off. Then it works also, just not sure if it might attach something internal additionally or not.

> # Azure services instrumentation guide

Reviewer: ```suggestion
# Azure instrumentation guide
```

to match with main `README.md`

## App Service

1. Choose your App Service.
2. Navigate to `Development Tools` > `Extensions`.
3. Find and install the `SignalFx .NET Tracing` extension.
4. Navigate to `Settings` > `Configuration`.

> 5. Add `New application setting`s to configur the receiver: 

Reviewer: ```suggestion
5. Add `New application setting`s to configure the receiver:
```

    ```
    Name: SIGNALFX_ACCESS_TOKEN 
    Value: (Your SIGNALFX access token)

    Name: SIGNALFX_ENDPOINT_URL
    Value: (Your Collector or SignalFX ingest endpoint)
    ```
    (See [advanced-config.md](advanced-config.md) for more options.)

> 6. Make sure that App Service was restarted.

Reviewer: maybe simply:

```suggestion
6. Restart the App Service.
```

# Azure services instrumentation guide

## App Service

1. Choose your App Service.
2. Navigate to `Development Tools` > `Extensions`.
3. Find and install the `SignalFx .NET Tracing` extension.
4. Navigate to `Settings` > `Configuration`.
5. Add `New application setting`s to configur the receiver: 

    ```
    Name: SIGNALFX_ACCESS_TOKEN 
    Value: (Your SIGNALFX access token)

    Name: SIGNALFX_ENDPOINT_URL
    Value: (Your Collector or SignalFX ingest endpoint)
    ```
    (See [advanced-config.md](advanced-config.md) for more options.)
6. Make sure that App Service was restarted.

> ## WebJobs (Experimental)

Author: Marked as experimental for now... hopefully we will get some feedback once someone tries experimental stuff.

Reviewer 3: I think it is better to record this info under internal. I'm not sure if there is demand for this capability.

Reviewer 2: I agree, we have supported only App Serivices. As long as there is not such request we should put your documentation into internal-doc/jira tasks to write document for WebJobs.

It is good candidate for upstram sync phase 3.

    ## version must match SignalFx .NET Tracing extension version
    Name: CORECLR_PROFILER_PATH_32
    Value: C:\home\signalfx\tracing\v0.2.0\x86\SignalFx.Tracing.ClrProfiler.Native.dll

    ## version must match SignalFx .NET Tracing extension version
    Name: CORECLR_PROFILER_PATH_64
    Value: C:\home\signalfx\tracing\v0.2.0\x64\SignalFx.Tracing.ClrProfiler.Native.dll

    ## General variables

    ## version must match SignalFx .NET Tracing extension version
    Name: SIGNALFX_DOTNET_TRACER_HOME
    Value: C:\home\signalfx\tracing\v0.2.0

    Name: SIGNALFX_PROFILER_EXCLUDE_PROCESSES
    Value: SnapshotUploader.exe;workerforwarder.exe

    ## version must match SignalFx .NET Tracing extension version
    Name: SIGNALFX_TRACE_LOG_PATH
    Value: C:\home\LogFiles\signalfx\tracing\v0.2.0\dotnet-profiler.log

    ## necessary to enable background services instrumentation
    Name: SIGNALFX_AZURE_APP_SERVICES
    Value: 0    

    Name: SIGNALFX_ACCESS_TOKEN 
    Value: (Your SIGNALFX access token)

    Name: SIGNALFX_ENDPOINT_URL
    Value: (Your Collector or SignalFX ingest endpoint)
    ```
 
## Azure Functions

Currently not supported. 

> <!-- TODO: What are the workarounds we can provide? -->

Author: Anything we can suggest here? OTel.NET SDK seems to work here, but we don't have it vendored yet.

Author: @pjanotti @pellared ⬆️

Reviewer 3: For AZ functions we can do https://github.com/signalfx/signalfx-dotnet-tracing#azure-function-custom-linux-docker-image

IIRC there is no way to install a profiler for Windows based functions.

Reviewer: I suggest postponing this (create a Jira story)

btw you can also find some dockerfiles here: https://github.com/signalfx/signalfx-dotnet-tracing/tree/next-ver/tracer/samples/ConsoleApp

2. Navigate to `Development Tools` > `Extensions`.
3. Find and install the `SignalFx .NET Tracing` extension.
4. Navigate to `Settings` > `Configuration`.

> 5. Add `New application setting`s to configure receiver: 

Reviewer 4: ```suggestion
5. Add `New application settings` to configure the receiver: 
```

Author: The button name is actually in the singular, that's why the singular form is highlighted. Maybe just to rephrase entirely differently?

Reviewer: ```suggestion
5. Add `New application setting` to configure the receiver: 
```

?

    (See [advanced-config.md](advanced-config.md) for more options.)
6. Make sure that App Service was restarted.

## WebJobs (Experimental)

> Repeat all of the App Service section steps. Currently environment configuration must be done manually. Specially notice that `SIGNALFX_AZURE_APP_SERVICES` must be disabled. It is strongly adviced to have a separate App Service for the Webjobs.

Reviewer 3: I think that the right way to do this is to disable SIGNALFX_AZURE_APP_SERVICES for the scope of the WebJob process and call attention that if you do it for the whole box you will lose the instrumentation for the Web App.
